### PR TITLE
Adaptable Dashboard Tabs

### PR DESCRIPTION
**Pull request by kyleknighted**
## Description

- Moves tabs to a Dropdown depending on how wide the browser is
- If hidden tab is chosen, it is locked to the end of available tabs to show it's active

Known issues:
- Adding a new Tab is seemingly random, so initial addition will lock to the end, but after a new Tab is selected, it will go back to it's random position. Potentially resolvable from API on the way it returns data
- Rapid continual browser resizing can cause tabs to not take 100% width of the available area, but seems to be a very edge-case and potentially not a real world problem

## Motivation and Context

Too many tabs on a Dashboard leads to tabs wrapping to the next line and generally a poor experience and visual representation of our brand

## Screenshots (if appropriate):
image
image
image
image
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by kmerz — APPROVED** [review]
LGTM :+1: Worked on my machine.

File: graylog2-web-interface/src/views/components/QueryTabs.tsx [hunk 1]
```diff
@@ -15,128 +15,204 @@
...
-  static propTypes = {
-    children: PropTypes.node,
-    onRemove: PropTypes.func.isRequired,
-    onSelect: PropTypes.func.isRequired,
-    onTitleChange: PropTypes.func.isRequired,
-    queries: PropTypes.object.isRequired,
-    selectedQueryId: PropTypes.string.isRequired,
-    titles: PropTypes.object.isRequired,
-  }
+  let maxWidth = moreBtn.offsetWidth + newBtn.offsetWidth + 30; // magic number is PageContentLayout__Container padding
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/QueryTabs.tsx:101]: I looks like we do not need to consider the page content layout padding here.

**Review by linuspahl — COMMENTED** [review]
Code looks good to me and it works well, I like the new page title styling.

The only suggestion I have, let's display background color for the active page title. We could use the same background color for the three dots, if one of the hidden pages is active. I saw that we are already adjusting the dots color, but this way it is easier to understand that one of the hidden pages is active.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Testing it fails for me with this, when creating/opening a dashboard:

```
Uncaught DOMException: Failed to execute 'querySelectorAll' on 'Element': 'li:not(.dropdown):not(.query-tabs-new):not(.dropdown-menu li)' is not a valid selector.
    at adjustTabs (http://localhost:8080/app.js:66888:32)
```

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
I got it working on Firefox and tested it a little bit. It looks better, but it also requires a second click when going to a later tab. 

Important: When a tab is active, we should not hide its title. When a tab is selected from the ellipsis menu, the title is hidden right now.

File: graylog2-web-interface/src/views/components/QueryTabs.tsx [hunk 2]
```diff
@@ -15,128 +15,241 @@
...
+             className="query-tabs-new">
+      <Icon name="plus" />
+    </NavItem>
+  );
+
+  const dropDownTabs = queryTabs('menuItem');
+
+  useEffect(() => {
+    adjustTabs();
+    window.addEventListener('resize', throttle(adjustTabs, 250));
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/QueryTabs.tsx:208]: A potentially easier solution would also be to use the `SizeMe` component from `react-sizeme` instead, which could potentially also simplify the logic used to generate the tabs. At least we can hand off the size provisioning and handling of window resizes to it.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
One remark regarding size calculation, also: can you write some tests for it?

**Review by linuspahl — APPROVED** [review]
Looks good to me!

**Comment by kmerz** [comment]
I tested the version again. And beside the known issues I could not see any problems. :heavy_check_mark:

**Review by dennisoelkers — APPROVED** [review]
Works greats and looks good now! Thanks for the patience of everyone involved! :heart:

**Pull request merged**